kv/memdb/memdb_test.go

```diff
@@ -249,90 +364,157 @@ func (s testMemDBSuite) TestRandom(c *C) {
-		case 0, 1:
-			newValue := make([]byte, rand.Intn(19)+1)
-			rand.Read(newValue)
-			p1.Put(k, newValue)
-			_ = p2.Put(k, newValue)
-		case 2:
-			p1.Delete(k)
-			_ = p2.Delete(k)
-		}
-	}
+	check := func() {
+		c.Check(p1.Len(), Equals, p2.Len())
+		c.Check(p1.Size(), Equals, p2.Size())
 
-	c.Check(p1.Len(), Equals, p2.Len())
-	c.Check(p1.Size(), Equals, p2.Size())
+		it1 := p1.NewIterator()
+		it1.SeekToFirst()
 
-	it1 := p1.NewIterator()
-	it1.SeekToFirst()
+		it2 := p2.NewIterator(nil)
 
-	it2 := p2.NewIterator(nil)
+		var prevKey, prevVal []byte
+		for it2.First(); it2.Valid(); it2.Next() {
+			c.Check(it1.Key(), BytesEquals, it2.Key())
+			c.Check(it1.Value(), BytesEquals, it2.Value())
 
-	var prevKey, prevVal []byte
-	for it2.First(); it2.Valid(); it2.Next() {
-		c.Check(it1.Key(), BytesEquals, it2.Key())
-		c.Check(it1.Value(), BytesEquals, it2.Value())
+			it := p1.NewIterator()
+			it.Seek(it2.Key())
+			c.Check(it.Key(), BytesEquals, it2.Key())
+			c.Check(it.Value(), BytesEquals, it2.Value())
 
-		it := p1.NewIterator()
-		it.Seek(it2.Key())
-		c.Check(it.Key(), BytesEquals, it2.Key())
-		c.Check(it.Value(), BytesEquals, it2.Value())
+			it.SeekForPrev(it2.Key())
+			c.Check(it.Key(), BytesEquals, it2.Key())
+			c.Check(it.Value(), BytesEquals, it2.Value())
 
-		it.SeekForPrev(it2.Key())
-		c.Check(it.Key(), BytesEquals, it2.Key())
-		c.Check(it.Value(), BytesEquals, it2.Value())
+			if prevKey != nil {
+				it.SeekForExclusivePrev(it2.Key())
+				c.Check(it.Key(), BytesEquals, prevKey)
+				c.Check(it.Value(), BytesEquals, prevVal)
+			}
 
-		if prevKey != nil {
-			it.SeekForExclusivePrev(it2.Key())
-			c.Check(it.Key(), BytesEquals, prevKey)
-			c.Check(it.Value(), BytesEquals, prevVal)
+			it1.Next()
+			prevKey = it2.Key()
+			prevVal = it2.Value()
 		}
 
-		it1.Next()
-		prevKey = it2.Key()
-		prevVal = it2.Value()
+		it1.SeekToLast()
+		for it2.Last(); it2.Valid(); it2.Prev() {
+			c.Check(it1.Key(), BytesEquals, it2.Key())
+			c.Check(it1.Value(), BytesEquals, it2.Value())
+			it1.Prev()
+		}
 	}
 
-	it1.SeekToLast()
-	for it2.Last(); it2.Valid(); it2.Prev() {
-		c.Check(it1.Key(), BytesEquals, it2.Key())
-		c.Check(it1.Value(), BytesEquals, it2.Value())
-		it1.Prev()
+	for _, k := range keys {
+		newValue := make([]byte, rand.Intn(19)+1)
+		rand.Read(newValue)
+		p1.Put(k, newValue)
+		_ = p2.Put(k, newValue)
 	}
-}
+	check()
 
-func (s testMemDBSuite) fillDB(cnt int) *DB {
-	p := New(4 * 1024)
-	var buf [4]byte
-	for i := 0; i < cnt; i++ {
-		binary.BigEndian.PutUint32(buf[:], uint32(i))
-		p.Put(buf[:], buf[:])
+	// This case try to ensure `Sandbox.merge` can work properly on random keys.
+	for i := range keys {
+		if rand.Intn(4) <= 2 {
+			continue
+		}
+		keys[i] = make([]byte, rand.Intn(19)+1)
+		rand.Read(keys[i])
 	}
+
+	sb := p1.Derive()
+	for _, k := range keys {
+		newValue := make([]byte, rand.Intn(19)+1)
+		rand.Read(newValue)
+		sb.Put(k, newValue)
+		_ = p2.Put(k, newValue)
+	}
+	sb.Flush()
```

Comment:
We need to add random cases for `p1` Derive multiple times, which is the execution pattern for a transaction.
Each round of Derive can put subset of the old keys or some new keys.